src/main/java/com/epam/controller/TrainerController.java

```diff
@@ -0,0 +1,97 @@
+        String username = UserUtils.getBaseUsername(request.getFirstname(), request.getLastname());
+        Optional<Trainee> traineeOptional = traineeService.getTraineeByUsername(username);
+        if (traineeOptional.isPresent()) {
+            ResponseEntity.badRequest().body("Not possible to register as a trainer and trainee both");
+        }
+
+        Trainer trainer = new Trainer();
+        User user = new User();
+        user.setFirstname(request.getFirstname());
+        user.setLastname(request.getLastname());
+        trainer.setUser(user);
+        TrainingType trainingType = new TrainingType(request.getSpecialization(), List.of(trainer));
+        trainer.setSpecialization(trainingType);
+
+
+        Trainer createdTrainer = trainerService.createTrainer(trainer);
+        TrainerRegistrationResponse response = new TrainerRegistrationResponse(createdTrainer.getUser().getUsername(), createdTrainer.getUser().getPassword());
+        return ResponseEntity.ok(response);
+    }
+
+    @GetMapping
+    @Transactional
+    public ResponseEntity<TrainerProfileResponse> getTrainerProfile(@RequestParam String username) {
+        Optional<Trainer> trainerOptional = trainerService.getTrainerByUsername(username);
+        if (trainerOptional.isEmpty()) {
+            ResponseEntity.notFound();
+        }
+        Trainer trainer = trainerOptional.get();
+        TrainerProfileResponse profileResponse = mapper.trainerToTrainerProfileResponse(trainer);
+        return ResponseEntity.ok(profileResponse);
+    }
+
+    @PutMapping
+    public ResponseEntity<TrainerUpdateResponseDto> updateTrainerProfile(@RequestParam(name = "username") String username, @RequestBody TrainerUpdateRequestDto request) {
+        Optional<Trainer> updatedTrainerOptional = trainerService.getTrainerByUsername(username);
+        if (updatedTrainerOptional.isEmpty()) {
+            return ResponseEntity.notFound().build();
+        }
+        Trainer updatedTrainer = updatedTrainerOptional.get();
+        mapper.updateEntity(updatedTrainer, request);
+        TrainingType trainingType = new TrainingType(request.getSpecialization(), List.of(updatedTrainer));
+        updatedTrainer.setSpecialization(trainingType);
+        Trainer trainer = trainerService.updateTrainer(updatedTrainer).get();
+
+        TrainerUpdateResponseDto updateResponseDto = mapper.trainerToTrainerUpdateResponseDto(trainer);
+        return ResponseEntity.ok(updateResponseDto);
+    }
+
+    @GetMapping("/trainings")
+    @Transactional
+    public ResponseEntity<List<TrainerTrainingResponse>> getTrainerTrainings(@Valid TrainerWithTraining trainerWithTraining) {
+        List<Training> trainings = trainerService.getTrainerTrainingsList(trainerWithTraining);
+
+        return ResponseEntity.ok(mapper.trainingsToTrainerTrainingsResponse(trainings));
+    }
+
+    @PatchMapping("activate-deactivate")
```

Comment:
"/activate" and "/deactivate".
Please have these two. Even though these are the same from your code vision, from REST POV these are two different actions. And users will call them clearly that way, right now it looks confusing and did you miss a slash before "activate-deactivate"?

Reply:
You will have no boolean param here in the new implementation as well. Boolean params could be really confusing, try to avoid them as much as possible.
```java
ResponseEntity<String> activateTrainee(String username) {...}
ResponseEntity<String> deactivateTrainee(String username) {...}

//instead of
ResponseEntity<String> activateDeactivateTrainee(String username, boolean isActive) {...}
```